Introduce a new migrate process instance rpc

berkaycanbc opened the pull request:
## Description

<!-- Please explain the changes you made here. -->

No tests are added for response because we do not put data to the migration response, so nothing to assert.

## Related issues

<!-- Which issues are closed by this PR or are related -->

closes #15119 

<!-- Cut-off marker
_All lines under and including the cut-off marker will be removed from the merge commit message_

## Definition of Ready

Please check the items that apply, before requesting a review.

You can find more details about these items in our wiki page about [Pull Requests and Code Reviews](https://github.com/camunda/zeebe/wiki/Pull-Requests-and-Code-Reviews).

* [ ] I've reviewed my own code
* [ ] I've written a clear changelist description
* [ ] I've narrowly scoped my changes
* [ ] I've separated structural from behavioural changes
-->

## Definition of Done

<!-- Please check the items that apply, before merging or (if possible) before requesting a review. -->

_Not all items need to be done depending on the issue and the pull request._

Code changes:
* [ ] The changes are backwards compatibility with previous versions
* [ ] If it fixes a bug then PRs are created to [backport](https://github.com/camunda/zeebe/compare/stable/0.24...main?expand=1&template=backport_template.md&title=[Backport%200.24]) the fix to the last two minor versions. You can trigger a backport by assigning labels (e.g. `backport stable/1.3`) to the PR, in case that fails you need to create backports manually.

Testing:
* [x] There are unit/integration tests that verify all acceptance criterias of the issue
* [ ] New tests are written to ensure backwards compatibility with further versions
* [ ] The behavior is tested manually
* [ ] The change has been verified by a QA run
* [ ] The impact of the changes is verified by a benchmark

Documentation:
* [ ] The documentation is updated (e.g. BPMN reference, configuration, examples, get-started guides, etc.)
* [ ] If the PR changes how BPMN processes are validated (e.g. support new BPMN element) then the Camunda modeling team should be informed to adjust the BPMN linting.

Other teams:
If the change impacts another team an issue has been created for this team, explaining what they need to do to support this change.
- [ ] [Operate](https://github.com/camunda/operate/issues)
- [ ] [Tasklist](https://github.com/camunda/tasklist/issues)
- [ ] [Web Modeler](https://github.com/camunda/web-modeler/issues)
- [ ] [Desktop Modeler](https://github.com/camunda/camunda-modeler/issues)
- [ ] [Optimize](https://github.com/camunda/camunda-optimize/issues)

Please refer to our [review guidelines](https://github.com/camunda/zeebe/wiki/Pull-Requests-and-Code-Reviews#code-review-guidelines).

korthout reviewed (changes requested):
Nice work @berkaycanbc 🏗️ 

❌ There are only two things that require change.

🔧 I've also made some suggestions to improve the wording of the documentation.

👍 The implementation and test look clean and were easy to review

berkaycanbc commented on gateway-protocol/src/main/proto/gateway.proto:
```diff
@@ -888,6 +913,28 @@ service Gateway {
 
+  /*
+    Migrates the process instance to the specified process definition.
+    In simple terms, this is handled by updating the active element's process and scope keys.
```
❓ Won't we do that for service task migrations?

korthout replied:
🙇 Oops, seems I misunderstood what you meant by scope keys. 

👍 You're right that the flow scope key of an element instance can be updated.

berkaycanbc commented:
Great feedback @korthout! Very good updates to make our documentation more explicit!

I sent changes, please take a look. 👍

korthout commented on gateway-protocol/src/main/proto/gateway.proto:
```diff
@@ -915,21 +915,23 @@ service Gateway {
-        - mapping instructions cannot change the element type (or event type)
-        - mapping instructions must refer to supported elements (i.e. some elements will be supported later on)
-        - mapping instructions must refer to elements in supported scenarios.
+        - not all active elements in the source process instance are mapped to the elements in the target process
+        - mapping instructions doesn't map the same type of elements or events
+        - mapping instructions doesn't refer to supported elements (i.e. some elements will be supported later on)
+        - mapping instructions doesn't refer to elements in supported scenarios.
```
🔧 Typically it's easier to understand statements when they are in a positive form rather than a negative.

E.g. `a mapping instruction changes the type of an element or event` can be considered easier to understand than `a mapping instruction doesn't map the same type of elements or events`

berkaycanbc replied:
That is actually true. Updated! 👍

korthout reviewed (approved):
Nice work @berkaycanbc 👍 

🔧 I have one last minor suggestion

LGTM 👍

berkaycanbc commented:
With PR #15230, we now need to define gateway service config for endpoints. I updated the config file and add MigrateProcessInstance.

@korthout could you please review my last commit for that? I also added a comment to the commit that explains the reasoning behind config selection.

korthout commented on gateway-protocol/src/main/proto/gateway.proto:
```diff
@@ -888,6 +913,30 @@ service Gateway {
+      FAILED_PRECONDITION:
+        - not all active elements in the source process instance are mapped to the elements in the target process
+        - a mapping instruction changes the type of an element or event
+        - a mapping instruction refers to an supported element (i.e. some elements will be supported later on)
```
🔧 typo: should be `unsupported element`

berkaycanbc replied:
🤦 updated!

korthout reviewed (commented):
Thanks @berkaycanbc 

❓ I wonder whether the migrate process instance rpc is actually idempotent. The process instance will be on the target process definition after the first successful (but timed out) request. The retried request is redundant because the process instance is already on the target process definition. We can simply respond to the second request with a successful MIGRATED event. What do you think?

berkaycanbc commented:
> Thanks @berkaycanbc
> 
> ❓ I wonder whether the migrate process instance rpc is actually idempotent. The process instance will be on the target process definition after the first successful (but timed out) request. The retried request is redundant because the process instance is already on the target process definition. We can simply respond to the second request with a successful MIGRATED event. What do you think?

How are we going to handle that then? Are we going to check if process instance has the target process definition? Then, how are we going to make sure it is **actually** migrated. We still need to observe events, right?

@korthout

korthout commented:
>How are we going to handle that then? Are we going to check if process instance has the target process definition? Then, how are we going to make sure it is actually migrated. We still need to observe events, right?

@berkaycanbc if process instance has the target process definition then there is nothing more to do. The process instance is already on that process definition. We can either:
- append all `ELEMENT_MIGRATED` and `MIGRATED` events again, even though they don't change any data
- only append the `ProcessInstanceMigration:MIGRATED` event to indicate that migration was successful.

berkaycanbc commented:
> > How are we going to handle that then? Are we going to check if process instance has the target process definition? Then, how are we going to make sure it is actually migrated. We still need to observe events, right?
> 
> @berkaycanbc if process instance has the target process definition then there is nothing more to do. The process instance is already on that process definition. We can either:
> 
> * append all `ELEMENT_MIGRATED` and `MIGRATED` events again, even though they don't change any data
> * only append the `ProcessInstanceMigration:MIGRATED` event to indicate that migration was successful.

That means, we will not reject the request if a `sourceElementId` doesn't exist in given process instance (which can happen successful but timeout requests). We will first check if given process instance has the target process definition id; if yes migration was successful, if not we will try to migrate.

TBH, I am not sure if that behaviour makes sense because it can lead to misunderstanding of the concept of process instance migration. WDYT? @korthout

korthout commented:
There's a third option: reject the migration if it's already on that version. That could be considered idempotent as well, as the resulting state is the same. Perhaps that has a better UX because it avoids this confusion you mention. We can clarify why it is rejected with a clear message

berkaycanbc commented:
> There's a third option: reject the migration if it's already on that version. That could be considered idempotent as well, as the resulting state is the same. Perhaps that has a better UX because it avoids this confusion you mention. We can clarify why it is rejected with a clear message

I think, this should work. Also, I agree it is a better UX since when user sees "you cannot migrate process instance to the same process definition it has", they will get the sense that it is already migrated.

Then, shall I move the endpoint to the retriable on `DEADLINE_EXCEEDED` side in the gateway service config? @korthout

berkaycanbc commented:
bors merge
